Enable systemd allow nonroot

sallyom opened the pull request:
This PR 
* adds a simple image, to contain a script that calls systemctl commands on the host
* adds the function to systemd method to enable services, root user
* adds message for non-root users to run `systemctl --user enable service --now` on host to enable service as non-root
        * the non-root systemd startup is blocked at this time, need more work with the permissions/mounts/etc
TODO on follow-up:
* add systemd flow to stop & restart (update) services
* add logic for non-root enable

cooktheryan commented on images/systemd/Dockerfile-systemd-enable:
```diff
@@ -0,0 +1,4 @@
+FROM registry.access.redhat.com/ubi8/ubi:latest
```
@sallyom we can probably drop these in `method_containers`

cooktheryan commented on pkg/engine/api/types.go:
```diff
@@ -29,10 +29,15 @@ type Raw struct {
-	InitialRun bool
-	LastCommit *object.Commit
+	Root       bool   `mapstructure:"root"`
+	// If non-root, home directory must be supplied to place unit file at $HOME/.config/systemd/user
```
@sallyom should we do the equivalent of a mkdir -p just to ensure the path exists? I know its extra work on us but I dont think the subdirectory exists by default

cooktheryan replied:
we may also have to do a chown as well or set permissions to be 777

cooktheryan replied:
that would be just for the non root users

sallyom replied:
I _think_ in order for homeDir to be non-root, host podman would be non-root socket, which means that directory will exist? I see `~/.config/systemd/user/sockets.target.wants/podman.socket` on my system and am assuming the act of `systemctl --user enable podman.socket` creates that if it doesn't exist. wdyt?

cooktheryan replied:
Ah yes good call.

sallyom commented on examples/systemd-non-root-enable.yaml:
```diff
@@ -0,0 +1,11 @@
+  systemd:
+    targetPath: examples/systemd
+    root: false
+    nonRootHomeDir: "/home/somalley"
```
* will update this with the nonroot-homedir of the ubuntu instance

sallyom replied:
@cooktheryan what's the non-root homedir of the ubuntu instance in gh actions? :)

cooktheryan commented:
@sallyom do you want to create the GH actions for building the image, archive, then load into the systemd test?

sallyom commented:
> @sallyom do you want to create the GH actions for building the image, archive, then load into the systemd test?

I added the systemd-enable test, is that what you mean here?

sallyom commented:
> > @sallyom do you want to create the GH actions for building the image, archive, then load into the systemd test?
> 
> I added the systemd-enable test, is that what you mean here? edit: you mean the new systemd-script! got it, will do that - we should wait to merge this til that is in, probably

Added the cross-build targets for `quay.io/harpoon/harpoon-systemd:latest` 
(`quay.io/harpoon/harpoon-systemd-amd:latest`, `quay.io/harpoon/harpoon-systemd-arm:latest`)

cooktheryan commented:
perfect that was exactly what i meant :D

cooktheryan commented on .github/workflows/docker-image.yml:
```diff
@@ -265,7 +312,46 @@ jobs:
+
+      - name: tag the image
+        run: sudo podman tag quay.io/harpoon/harpoon-amd:latest quay.io/harpoon/harpoon:latest
+
```
we will need to do the same steps of loading the systemd image and tagging it as harpoon-systemd:latest here that way we test against the newly created image

cooktheryan replied:
that should allow for us to pass the failing test

cooktheryan commented on pkg/engine/systemd.go:
```diff
@@ -12,7 +17,57 @@ func systemdPodman(ctx context.Context, mo *FileMountOptions) error {
+	os.Setenv("SERVICE", service)
+	s := specgen.NewSpecGenerator(systemdImage, false)
+	runMount := "/run/systemd"
+	if !root {
```
We may need an enhancement request for figuring out the uid dynamically for the user but it should not stop this PR from merging
